src/Encoding.chpl

```diff
@@ -0,0 +1,248 @@
+/* Documentation for Encoding */
+module Encoding {
+
+  use IO;
+
+  type writingChannel = channel(true,iokind.little,false);
+  type readingChannel = channel(false,iokind.little,false);
+
+  // wireTypes
+  const varint = 0;
+  const lengthDelimited = 2;
+  const fixed64Type = 1;
+  const fixed32Type = 5;
+
+  proc unsignedVarintAppend(val:uint, ch: writingChannel) throws {
+    if val == 0 {
+      var zero: uint(8);
+      ch.write(zero);
+      return;
+    }
+  
+    var newVal = val;
+    var shiftVal: uint;
+    while newVal != 0 {
+      shiftVal = newVal >> 7;
+      var k = if shiftVal != 0 then 0x80 else 0x00;
+      var newByte = (newVal & 0x7F | k):uint(8);
+      ch.write(newByte);
+      newVal = shiftVal;
+    }
+  }
+
+  proc unsignedVarintConsume(ch: readingChannel): (uint, int) throws {
+    var shift = 0;
+    var val:uint;
+    var len = 0;
+    var s:uint(8);
+    while true {
+      if !ch.read(s) then return (val, -1);
+      val = val + ((s & 0x7F): uint << shift);
+      shift = shift + 7;
+      len = len + 1;
+      if (s & 0x80) == 0 then break;
+    }
+    return (val, len);
+  }
+
+  proc tagConsume(ch: readingChannel) throws {
+    var (tag, tlen) = unsignedVarintConsume(ch);
+    if tlen == -1 then return -1;
+    var wireType = (tag & 0x7): int;
```

Comment:
this and many other `var wireType` or `const wireTiype` are unused variables.

That might mean you can remove the `wireTypes` consts at the top, too.

Reply:
Hmm..I would rather prefer keeping these for now, there are used when we ensure backward compatibility, though we can remove them anytime if they turn out to be useless.

Reply:
Sure, it makes sense to keep them in this PR if you plan to use them in a future PR.